From 0ccd30b588c6824f87f3cc7610d31ecb8e6e156a Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Wed, 16 Apr 2025 18:13:47 +0200
Subject: [PATCH] Issue 6736 - Exception thrown by dsconf instance repl get_ruv
 (#6742)

After issue #6715 commit dsconf instance replication get_ruv fails with error
because by default NormalizedRidDict does not calling getitem as I expected.
The solution is to overwrite the get function to do things properly.

Issue: #6736
Relates: #6715

Reviewed by: @droideck (Thanks!)

(cherry picked from commit 64b2514d09b642913f2781ebdf8a5931f9e74eb6)
---
 .../suites/replication/regression_m2_test.py  | 42 +++++++++++++++++++
 src/lib389/lib389/replica.py                  | 12 +++++-
 2 files changed, 53 insertions(+), 1 deletion(-)

diff --git a/dirsrvtests/tests/suites/replication/regression_m2_test.py b/dirsrvtests/tests/suites/replication/regression_m2_test.py
index 2283b9041..ba1ffcc9c 100644
--- a/dirsrvtests/tests/suites/replication/regression_m2_test.py
+++ b/dirsrvtests/tests/suites/replication/regression_m2_test.py
@@ -1275,6 +1275,48 @@ def test_normalized_rid_dict():
         assert nrd[nkey] == val
 
 
+def test_get_with_normalized_rid_dict():
+    """Check lib389.replica NormalizedRidDict.get() function
+
+    :id: 4422e1be-1619-11f0-a37a-482ae39447e5
+    :setup: None
+    :steps:
+        1. Initialize a NormalizedRidDict
+        2. Check that normalization do something
+        3. Check that get() returns the expected value
+        4. Check get() with wrong key
+    :expectedresults:
+        1. Success
+        2. Success
+        3. Success
+        4. Should return the default value if it is provided
+           otherwise it should return None
+    """
+
+    sd = { '1': 'v1', '020': 'v2' }
+    nsd = { NormalizedRidDict.normalize_rid(key): val for key,val in sd.items() }
+    nkeys = list(nsd.keys())
+
+    # Initialize a NormalizedRidDict
+    nrd = NormalizedRidDict()
+    for key,val in sd.items():
+        nrd[key] = val
+
+    # Check that get() returns the expected value
+    for key,val in sd.items():
+        nkey = NormalizedRidDict.normalize_rid(key)
+        assert nrd.get(key) == val
+        assert nrd.get(nkey) == val
+        assert nrd.get(key, 'foo') == val
+        assert nrd.get(nkey, 'foo') == val
+
+    # Check get() with wrong key
+    assert nrd.get('99', 'foo2') == 'foo2'
+    assert nrd.get('099', 'foo2') == 'foo2'
+    assert nrd.get('99') is None
+    assert nrd.get('099') is None
+
+
 @pytest.mark.ds49915
 @pytest.mark.bz1626375
 def test_online_reinit_may_hang(topo_with_sigkill):
diff --git a/src/lib389/lib389/replica.py b/src/lib389/lib389/replica.py
index 0dc3a0f54..5183c41e9 100644
--- a/src/lib389/lib389/replica.py
+++ b/src/lib389/lib389/replica.py
@@ -841,6 +841,12 @@ class NormalizedRidDict(dict):
         nkey = NormalizedRidDict.normalize_rid(key)
         super().__setitem__(nkey, value)
 
+    def get(self, key, vdef=None):
+        try:
+            return self[key]
+        except KeyError:
+            return vdef
+
 
 class RUV(object):
     """Represents the server in memory RUV object. The RUV contains each
@@ -911,7 +917,7 @@ class RUV(object):
             ValueError("Wrong CSN value was supplied")
 
         timestamp = int(csn[:8], 16)
-        time_str = datetime.datetime.utcfromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
+        time_str = datetime.datetime.fromtimestamp(timestamp, datetime.UTC).strftime('%Y-%m-%d %H:%M:%S')
         # We are parsing shorter CSN which contains only timestamp
         if len(csn) == 8:
             return time_str
@@ -987,6 +993,10 @@ class RUV(object):
                 return False
         return True
 
+    def __str__(self):
+        return str(self.format_ruv())
+
+
 
 class ChangelogLDIF(object):
     def __init__(self, file_path, output_file):
-- 
2.49.0

